src/include/miopen/memref.hpp

```diff
@@ -0,0 +1,82 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2021 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+namespace miopen {
+
+// clang-format off
+template <int N>
+void dropFront(int64_t arr[N], int64_t* res)
+{
+  for (unsigned i = 1; i < N; ++i)
+    *(res + i - 1) = arr[i];
+}
+
+//===----------------------------------------------------------------------===//
+// Codegen-compatible structures for StridedMemRef type.
+//===----------------------------------------------------------------------===//
+/// StridedMemRef descriptor type with static rank.
+template <typename T, int N>
+struct StridedMemRefType {
+  T* basePtr;
+  T* data;
+  int64_t offset;
+  int64_t sizes[N];
+  int64_t strides[N];
```

Comment:
> When/if the implementation of StridedMemRefType<> in MLIR change, you would need to update that struct type in Miir.h.

Would you be pulling in this type somewhere else into MIOpen, like an existing PR [here](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/631)?

Reply:
> you would need to update that struct type in `Miir.h`.

I agree about exporting that in the library side. I guess when you later decided that MIOpen also need to pull it in, it can be done in a different namespace.

Reply:
@atamazov One drawback is that if we are to export this in the C header, I will have to rewrite it in C format since this is template based. But on the other hand, I really want to avoid touching this because it make any changes in this type easy to observe in MLIR source.